cvat/apps/dataset_manager/annotation.py

```diff
@@ -442,34 +443,296 @@ def normalize_shape(shape):
+                    interpolatedPoints.append({
+                        "x": leftPoint["x"] + (rightPoint["x"] - leftPoint["x"]) * offset,
+                        "y": leftPoint["y"] + (rightPoint["y"] - leftPoint["y"]) * offset,
+                    })
+
+            reducedPoints = reduceInterpolation(
+                interpolatedPoints,
+                completedMatching,
+                leftPoints,
+                rightPoints,
+            )
+
+            return to_array(reducedPoints).tolist()
+
+        def polyshape_interpolation(shape0, shape1):
             shapes = []
-            is_same_type = shape0["type"] == shape1["type"]
             is_polygon = shape0["type"] == ShapeType.POLYGON
-            is_polyline = shape0["type"] == ShapeType.POLYLINE
-            is_same_size = len(shape0["points"]) == len(shape1["points"])
-            if not is_same_type or is_polygon or is_polyline or not is_same_size:
-                shape0 = TrackManager.normalize_shape(shape0)
-                shape1 = TrackManager.normalize_shape(shape1)
+            if is_polygon:
+                shape0["points"].extend(shape0["points"][:2])
+                shape1["points"].extend(shape1["points"][:2])
 
             distance = shape1["frame"] - shape0["frame"]
-            step = np.subtract(shape1["points"], shape0["points"]) / distance
             for frame in range(shape0["frame"] + 1, shape1["frame"]):
-                off = frame - shape0["frame"]
+                offset = (frame - shape0["frame"]) / distance
+                points = None
                 if shape1["outside"]:
-                    points = np.asarray(shape0["points"]).reshape(-1, 2)
-                else:
-                    points = (shape0["points"] + step * off).reshape(-1, 2)
-                shape = deepcopy(shape0)
-                if len(points) == 1:
-                    shape["points"] = points.flatten()
+                    points = np.asarray(shape0["points"])
                 else:
-                    broken_line = geometry.LineString(points).simplify(0.05, False)
-                    shape["points"] = [x for p in broken_line.coords for x in p]
+                    points = interpolate_position(shape0, shape1, offset)
+
+                shapes.append(copy_shape(shape0, frame, points))
+
+            if is_polygon:
+                shape0["points"] = shape0["points"][:-2]
+                shape1["points"] = shape1["points"][:-2]
+                for shape in shapes:
+                    shape["points"] = shape["points"][:-2]
+
+            return shapes
+
+        def interpolate(shape0, shape1):
+            is_same_type = shape0["type"] == shape1["type"]
+            is_rectangle = shape0["type"] == ShapeType.RECTANGLE
+            is_cuboid = shape0["type"] == ShapeType.CUBOID
+            is_polygon = shape0["type"] == ShapeType.POLYGON
+            is_polyline = shape0["type"] == ShapeType.POLYLINE
+            is_points = shape0["type"] == ShapeType.POINTS
```

Comment:
May be an exception should be thrown if type is anything else.

Reply:
@zhiltsov-max 

Yes, we can do it with ``if..elif`` structure below